File: src/ui/result_panel.rs
use crate::app::state::AppState;
use crate::domain::ConverterType;
use egui::{Color32, RichText, Ui};

/// Display calculated results in the right panel.
pub fn show_result_panel(ui: &mut Ui, state: &mut AppState) {
    let res = &state.results;

    ui.heading("📊 Results");
    ui.separator();

    // Show duty cycle for Buck/Boost
    match state.active_converter {
        ConverterType::Buck | ConverterType::Boost => {
            show_value_row(ui, "Duty cycle", state.params.duty_cycle, "");
        }
        _ => {}
    }
    show_value_row(ui, "Switching freq", state.params.frequency, "Hz");

    ui.add_space(4.0);
    ui.label(RichText::new("Voltage / Current").strong().size(14.0));
    ui.separator();
    show_value_row(ui, "Vout", res.vout, "V");
    show_value_row(ui, "Iout", res.iout, "A");
    show_value_row(ui, "Iin", res.iin, "A");

    // Show VSI-specific results
    if let Some(vrms) = res.rms_output {
        show_value_row(ui, "Vrms", vrms, "V");
    }
    if let Some(v1) = res.fundamental_amplitude {
        show_value_row(ui, "V1 (fund)", v1, "V");
    }

    ui.add_space(8.0);
    ui.label(RichText::new("Ripple").strong().size(14.0));
    ui.separator();
    show_value_row(ui, "V ripple (pp)", res.vout_ripple, "V");
    show_value_row(ui, "I_L ripple (pp)", res.il_ripple, "A");

    ui.add_space(8.0);
    ui.label(RichText::new("Losses & Efficiency").strong().size(14.0));
    ui.separator();
    show_value_row(ui, "Conduction loss", res.conduction_losses, "W");
    show_value_row(ui, "Switching loss", res.switching_losses, "W");
    show_value_row(
        ui,
        "Total loss",
        res.conduction_losses + res.switching_losses,
        "W",
    );

    // Efficiency with color coding
    let eff = res.efficiency;
    let eff_color = if eff > 0.95 {
        Color32::GREEN
    } else if eff > 0.85 {
        Color32::YELLOW
    } else {
        Color32::RED
    };

    ui.horizontal(|ui| {
        ui.label("Efficiency:");
        ui.label(
            RichText::new(format!("{:.1}%", eff * 100.0))
                .color(eff_color)
                .strong()
                .size(16.0),
        );
    });

    // THD for VSI
    if let Some(thd) = res.thd {
        ui.add_space(8.0);
        ui.label(RichText::new("Harmonics").strong().size(14.0));
        ui.separator();
        let thd_color = if thd < 0.5 {
            Color32::GREEN
        } else if thd < 1.0 {
            Color32::YELLOW
        } else {
            Color32::RED
        };
        ui.horizontal(|ui| {
            ui.label("THD:");
            ui.label(
                RichText::new(format!("{:.1}%", thd * 100.0))
                    .color(thd_color)
                    .strong(),
            );
        });
    }

    ui.add_space(12.0);
    ui.separator();
    ui.label(RichText::new("Status").strong().size(14.0));
    ui.label(&state.status_message);
}

fn show_value_row(ui: &mut Ui, label: &str, value: f64, unit: &str) {
    ui.horizontal(|ui| {
        ui.label(format!("{}:", label));
        ui.with_layout(egui::Layout::right_to_left(egui::Align::Center), |ui| {
            ui.colored_label(
                egui::Color32::from_rgb(150, 200, 255),
                format_value(value, unit),
            );
        });
    });
}

/// Format a value with appropriate SI prefix and significant digits.
fn format_value(value: f64, unit: &str) -> String {
    let abs_val = value.abs();

    if abs_val == 0.0 {
        return format!("0 {}", unit);
    }

    let (scaled, prefix) = if abs_val >= 1_000_000.0 {
        (value / 1_000_000.0, "M")
    } else if abs_val >= 1_000.0 {
        (value / 1_000.0, "k")
    } else if abs_val >= 1.0 {
        (value, "")
    } else if abs_val >= 0.001 {
        (value * 1_000.0, "m")
    } else if abs_val >= 0.000_001 {
        (value * 1_000_000.0, "μ")
    } else if abs_val >= 1e-9 {
        (value * 1e9, "n")
    } else {
        (value * 1e12, "p")
    };

    // Choose decimal places based on magnitude
    let decimals = if scaled.abs() > 100.0 {
        1
    } else if scaled.abs() > 10.0 {
        2
    } else {
        3
    };

    format!("{:.prec$} {}{}", scaled, prefix, unit, prec = decimals)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_value() {
        assert_eq!(format_value(12.0, "V"), "12.00 V");
        assert_eq!(format_value(0.005, "A"), "5.000 mA");
        assert_eq!(format_value(0.000_100, "H"), "100.00 μH");
        assert_eq!(format_value(1000.0, "Hz"), "1.000 kHz");
        assert_eq!(format_value(1_000_000.0, "Hz"), "1.000 MHz");
        assert_eq!(format_value(0.0, "V"), "0 V");
    }
}

File: src/ui/ResultPanel.cpp
#include "ResultPanel.h"
#include "utils/Formatting.h"

#include <wx/statbox.h>
#include <wx/statline.h>
#include <wx/colour.h>

ResultPanel::ResultPanel(wxWindow* parent, AppState& state)
    : wxPanel(parent, wxID_ANY)
    , m_state(state)
    , m_colors(ThemeColors::resolve(state.theme))
{
    SetBackgroundColour(m_colors.sidebar_bg);

    wxBoxSizer* mainSizer = new wxBoxSizer(wxVERTICAL);

    // ── Title ──
    wxStaticText* title = new wxStaticText(this, wxID_ANY, "   📊 Results");
    wxFont titleFont = title->GetFont();
    titleFont.SetPointSize(titleFont.GetPointSize() + 2);
    titleFont.SetWeight(wxFONTWEIGHT_BOLD);
    title->SetFont(titleFont);
    title->SetForegroundColour(m_colors.text_primary);
    mainSizer->Add(title, 0, wxEXPAND | wxALL, 8);
    mainSizer->Add(new wxStaticLine(this), 0, wxEXPAND | wxLEFT | wxRIGHT, 8);

    // ── Voltage / Current ──
    {
        wxStaticBoxSizer* box = new wxStaticBoxSizer(wxVERTICAL, this, "Voltage / Current");
        m_vout_text = new wxStaticText(this, wxID_ANY, "Vout = 0 V");
        m_iout_text = new wxStaticText(this, wxID_ANY, "Iout = 0 A");
        m_iin_text = new wxStaticText(this, wxID_ANY, "Iin = 0 A");
        m_vrms_text = new wxStaticText(this, wxID_ANY, "");
        m_v1_text = new wxStaticText(this, wxID_ANY, "");

        for (auto* t : {m_vout_text, m_iout_text, m_iin_text, m_vrms_text, m_v1_text})
            t->SetForegroundColour(m_colors.text_value);
        for (auto* t : {m_vout_text, m_iout_text, m_iin_text, m_vrms_text, m_v1_text})
            box->Add(t, 0, wxEXPAND | wxALL, 4);

        mainSizer->Add(box, 0, wxEXPAND | wxLEFT | wxRIGHT, 8);
    }

    mainSizer->AddSpacer(4);

    // ── Ripple ──
    {
        wxStaticBoxSizer* box = new wxStaticBoxSizer(wxVERTICAL, this, "Ripple");
        m_vout_ripple_text = new wxStaticText(this, wxID_ANY, "Vout ripple = 0 V");
        m_il_ripple_text = new wxStaticText(this, wxID_ANY, "iL ripple = 0 A");
        m_vout_ripple_text->SetForegroundColour(m_colors.text_value);
        m_il_ripple_text->SetForegroundColour(m_colors.text_value);
        box->Add(m_vout_ripple_text, 0, wxEXPAND | wxALL, 4);
        box->Add(m_il_ripple_text, 0, wxEXPAND | wxALL, 4);
        mainSizer->Add(box, 0, wxEXPAND | wxLEFT | wxRIGHT, 8);
    }

    mainSizer->AddSpacer(4);

    // ── Losses & Efficiency ──
    {
        wxStaticBoxSizer* box = new wxStaticBoxSizer(wxVERTICAL, this, "Losses & Efficiency");
        m_cond_loss_text = new wxStaticText(this, wxID_ANY, "Conduction = 0 W");
        m_sw_loss_text = new wxStaticText(this, wxID_ANY, "Switching = 0 W");
        m_total_loss_text = new wxStaticText(this, wxID_ANY, "Total = 0 W");
        m_efficiency_text = new wxStaticText(this, wxID_ANY, "Efficiency = 0%");

        for (auto* t : {m_cond_loss_text, m_sw_loss_text, m_total_loss_text, m_efficiency_text})
            t->SetForegroundColour(m_colors.text_value);
        for (auto* t : {m_cond_loss_text, m_sw_loss_text, m_total_loss_text, m_efficiency_text})
            box->Add(t, 0, wxEXPAND | wxALL, 4);

        mainSizer->Add(box, 0, wxEXPAND | wxLEFT | wxRIGHT, 8);
    }

    mainSizer->AddSpacer(4);

    // ── Harmonics (VSI) ──
    {
        wxStaticBoxSizer* box = new wxStaticBoxSizer(wxVERTICAL, this, "Harmonics");
        m_thd_text = new wxStaticText(this, wxID_ANY, "");
        m_thd_text->SetForegroundColour(m_colors.text_value);
        box->Add(m_thd_text, 0, wxEXPAND | wxALL, 4);
        mainSizer->Add(box, 0, wxEXPAND | wxLEFT | wxRIGHT, 8);
    }

    mainSizer->AddSpacer(8);
    mainSizer->Add(new wxStaticLine(this), 0, wxEXPAND | wxLEFT | wxRIGHT, 8);

    // ── Status ──
    m_status_text = new wxStaticText(this, wxID_ANY, "Status: ready");
    m_status_text->SetForegroundColour(m_colors.status);
    mainSizer->Add(m_status_text, 0, wxEXPAND | wxALL, 8);

    mainSizer->AddStretchSpacer();

    SetSizer(mainSizer);
    UpdateDisplay();
}

void ResultPanel::UpdateDisplay() {
    m_colors = ThemeColors::resolve(m_state.theme);
    SetBackgroundColour(m_colors.sidebar_bg);

    const auto& r = m_state.results;

    m_vout_text->SetLabel("Vout = " + formatting::format_value(r.vout, "V"));
    m_iout_text->SetLabel("Iout = " + formatting::format_value(r.iout, "A"));
    m_iin_text->SetLabel("Iin = " + formatting::format_value(r.iin, "A"));

    if (r.rms_output) {
        m_vrms_text->SetLabel("Vrms = " + formatting::format_value(*r.rms_output, "V"));
        m_vrms_text->Show();
    } else {
        m_vrms_text->Hide();
    }

    if (r.fundamental_amplitude) {
        m_v1_text->SetLabel("V1 (fund) = " + formatting::format_value(*r.fundamental_amplitude, "V"));
        m_v1_text->Show();
    } else {
        m_v1_text->Hide();
    }

    m_vout_ripple_text->SetLabel("Vout ripple = " + formatting::format_value(r.vout_ripple, "V"));
    m_il_ripple_text->SetLabel("iL ripple = " + formatting::format_value(r.il_ripple, "A"));

    m_cond_loss_text->SetLabel("Conduction = " + formatting::format_value(r.conduction_losses, "W"));
    m_sw_loss_text->SetLabel("Switching = " + formatting::format_value(r.switching_losses, "W"));
    m_total_loss_text->SetLabel("Total = " + formatting::format_value(r.conduction_losses + r.switching_losses, "W"));

    // Efficiency with color coding
    double eff = r.efficiency;
    wxColour eff_color;
    if (eff > 0.95)       eff_color = wxColour(0, 200, 0);    // Green
    else if (eff > 0.85) eff_color = wxColour(200, 200, 0);  // Yellow
    else                  eff_color = wxColour(200, 0, 0);    // Red

    m_efficiency_text->SetLabel(wxString::Format("Efficiency = %.1f%%", eff * 100.0));
    m_efficiency_text->SetForegroundColour(eff_color);

    // THD
    if (r.thd) {
        double thd_val = *r.thd;
        wxColour thd_color;
        if (thd_val < 0.5)       thd_color = wxColour(0, 200, 0);
        else if (thd_val < 1.0) thd_color = wxColour(200, 200, 0);
        else                     thd_color = wxColour(200, 0, 0);
        m_thd_text->SetLabel(wxString::Format("THD = %.1f%%", thd_val * 100.0));
        m_thd_text->SetForegroundColour(thd_color);
        m_thd_text->Show();
    } else {
        m_thd_text->Hide();
    }

    m_status_text->SetLabel("Status: " + m_state.status_message);
    m_status_text->SetForegroundColour(m_colors.status);

    // Refresh text colors for all visible labels
    for (wxWindow* child : GetChildren()) {
        wxStaticText* st = dynamic_cast<wxStaticText*>(child);
        if (st && st != m_efficiency_text && st != m_thd_text && st != m_status_text) {
            st->SetForegroundColour(m_colors.text_value);
        }
    }
}
